### PR TITLE
fix: Fix integer overflow in date_parser

### PR DESCRIPTION
**Pull request by eejbyfeldt**
## Which issue does this PR close?

This resolves one of the overflows discussed in #481

## Rationale for this change

The date_parser was introduced in #383 and is mostly a direct port of code in Spark. In Spark the code uses Scala which has defined integer overflow as two's complement wrapping. 

## What changes are included in this PR?

The proposed fixed is to use `std::num::Wrapping` to get the same wrapping behavior in rust.  The overflown value will still be rejected in a later check that uses `current_segment_digits` so allowing the overflow does not lead to correctness issues.

## How are these changes tested?

New and existing unit tests.

### CODE REVIEW

File: core/src/execution/datafusion/expressions/cast.rs [hunk 1]
```diff
@@ -1836,6 +1837,8 @@ mod tests {
             Some(" 202 "),
             Some("\n 2020-\r8 "),
             Some("2020-01-01T"),
+            // Overflows i32
+            Some("-4607172990231812908"),
```
Comment (andygrove) [inline core/src/execution/datafusion/expressions/cast.rs:1841]: ❤️

File: core/src/execution/datafusion/expressions/cast.rs [hunk 2]
```diff
@@ -19,6 +19,7 @@ use std::{
     any::Any,
     fmt::{Debug, Display, Formatter},
     hash::{Hash, Hasher},
+    num::Wrapping,
```
Comment (andygrove) [inline core/src/execution/datafusion/expressions/cast.rs:22]: I wasn't aware of `Wrapping`. TIL.

Comment (parthchandra) [reply core/src/execution/datafusion/expressions/cast.rs:22]: Wonder if this is also faster since there is no overflow check?

**Review by parthchandra — APPROVED** [review]

File: core/src/execution/datafusion/expressions/cast.rs (hunk 2 above)
Comment (viirya) [reply core/src/execution/datafusion/expressions/cast.rs:22]: I think we usually use some methods like `wrapping_add`, etc.

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @eejbyfeldt.

Thanks for the review @parthchandra

**Pull request merged**